File: 2022/13/13.py
import json
from functools import cmp_to_key

def compr(l, r):
  match [l, r]:
    case [None, _]: return -1
    case [_, None]: return 1
    case [[], []]: return 0
    case [[], _]: return -1
    case [_, []]: return 1
    case [a, b] if type(a) == int and type(b) == int: return 0 if a == b else [1, -1][a < b]
    case [a, b] if type(a) == int and type(b) == list: return compr([a], b)
    case [a, b] if type(a) == list and type(b) == int: return compr(a, [b])
    case [[a, *aa], [b, *bb]]: return compr(a, b) or compr(aa, bb)

packets = [json.loads(line) for line in open("input.txt").read().splitlines() if line != ""]
pairs = [[packets[i], packets[i+1]] for i in range(0, len(packets), 2)]

print("Part 1:", sum(i + 1 if compr(*pairs[i]) == -1 else 0 for i in range(len(pairs))))

packets.append([[2]])
packets.append([[6]])
packets.sort(key=cmp_to_key(compr))

print("Part 2:", (packets.index([[2]]) + 1) * (packets.index([[6]]) + 1))

File: 2022/13/13.clj
(ns aoc.2022.13.13 
  (:require [aoc.common :refer [find-index read-input sum]]
            [clojure.core.match :refer [match]]
            [clojure.edn :as edn]))

(defn compr [l r]
  (match [l r]
    [nil _] true
    [_ nil] false
    [[] []] nil
    [[] _] true
    [_ []] false
    [(a :guard int?) (b :guard int?)] (if (= a b) nil (< a b))
    [(a :guard int?) (b :guard vector?)] (compr [a] b)
    [(a :guard vector?) (b :guard int?)] (compr a [b])
    [[a & as] [b & bs]] (->> [(compr a b) (compr as bs)] (filter some?) first)))

(let [packets (->> (read-input :split-with #"\n+")
                   (map edn/read-string))
      dp1 [[2]]
      dp2 [[6]]
      sorted (->> (conj packets dp1 dp2)
                  (sort compr))]

  (->> packets
       (partition 2)
       (keep-indexed #(when (apply compr %2) %1))
       (map inc)
       sum
       (println "Part 1:"))

  (->> [(->> sorted (find-index #(= dp1 %2)) inc)
        (->> sorted (find-index #(= dp2 %2)) inc)]
       (apply *)
       (println "Part 2:")))